include/nds/arm9/math.h

```diff
@@ -71,7 +71,10 @@ extern "C" {
 ///     20.12 denominator.
 static inline void divf32_asynch(int32_t num, int32_t den)
 {
-    REG_DIV_NUMER = ((int64_t)num) << 12;
+    // Casting from uint64_t to int64_t is implementation defined, however
+    // it typically follows 2's complement behavior as is assumed here
+    // https://gcc.gnu.org/onlinedocs/gcc/Integers-implementation.html#Integers-implementation
+    REG_DIV_NUMER = (int64_t)((uint64_t)(int64_t)num << 12);
```

Comment:
I believe left shifting of negative values is undefined even if it doesn't affect the sign bit. I made this patch a couple weeks ago but if I remember correctly this line was triggering a crash when using runtime sanitizers in Clang/Zig.

From the latest C standard that I could find online (ISO/IEC 9899:2024, C23, open std n3220):

> The result of E1 << E2 is E1 left-shifted E2 bit positions; vacated bits are filled with zeros. If E1 has an unsigned type, the value of the result is E1 × 2E2, wrapped around. If E1 has a signed type and nonnegative value, and E1 × 2E2 is representable in the result type, then that is the resulting value; otherwise, the behavior is undefined.

I think the link might be useful to have in the code as documentation (since it's a notably odd series of casts) but then again, a `git blame` would show this anyways. Shall I move it to the commit message?

Reply:
Meh, that's annoying. Let's keep it in the code, then. That's not trivial. If we're going to have a complicated chain of casts we might as well have the explanation there... Git blame will keep it forever, but it will be harder to reach if the code moves around.